providers/jetstream/jsprovider.go

```diff
@@ -0,0 +1,367 @@
+	_, err = p.js.AccountInfo(context.Background())
+	if err != nil {
+		p.disconnect()
+		return err
+	}
+	p.kv, err = p.js.KeyValue(context.Background(), p.bucket)
+	if err != nil {
+		if err == jetstream.ErrBucketNotFound {
+			p.kv, err = p.js.CreateKeyValue(context.Background(), jetstream.KeyValueConfig{
+				Bucket: p.bucket,
+			})
+			if err != nil {
+				p.disconnect()
+				return err
+			}
+		} else if err != nil {
+			p.disconnect()
+			return err
+		}
+	}
+	return nil
+}
+
+func (p *Provider) disconnect() {
+	p.nc.Close()
+}
+
+// child entities are stored under <pk>.<childPublicKey>
+func (p *Provider) getChildren(prefix string) (map[string][]byte, error) {
+	entries, err := p.kv.History(context.Background(), fmt.Sprintf("%s.*", prefix))
+	if err != nil {
+		if err == jetstream.ErrKeyNotFound {
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	m := make(map[string][]byte)
+	for _, e := range entries {
+		if e.Operation() != jetstream.KeyValuePut {
+			delete(m, e.Key())
+			continue
+		} else {
+			n := e.Key()[len(prefix)+1:]
+			m[n] = e.Value()
+		}
+	}
+	return m, nil
+}
+
+func (p *Provider) Load() ([]*ab.OperatorData, error) {
+	m, err := p.getChildren(OperatorPrefix)
+	if err != nil {
+		return nil, err
+	}
+	operators := make([]*ab.OperatorData, 0, len(m))
+	for _, v := range m {
+		o := &ab.OperatorData{
+			BaseData: ab.BaseData{
+				Token: string(v),
+			},
+		}
+		oc, err := jwt.DecodeOperatorClaims(o.Token)
+		if err != nil {
+			return nil, err
+		}
+		o.Claim = oc
+		o.Loaded = o.Claim.IssuedAt
+		o.EntityName = o.Claim.Name
+		o.Key, err = p.getKey(o.Claim.Subject)
+		if err != nil {
+			return nil, err
+		}
+		for _, sk := range o.Claim.SigningKeys {
+			k, err := p.getKey(sk)
+			if err != nil {
+				return nil, err
+			}
+			o.OperatorSigningKeys = append(o.OperatorSigningKeys, k)
+		}
+		if err := p.loadAccounts(o); err != nil {
+			return nil, err
+		}
+		operators = append(operators, o)
+	}
+	return operators, nil
+}
+
+func (p *Provider) loadAccounts(od *ab.OperatorData) error {
+	// accounts stored under <operatorPublicKey>.<accountPublicKey>
+	m, err := p.getChildren(od.Claim.Subject)
+	if err != nil {
+		return err
+	}
+	for _, v := range m {
+		a := &ab.AccountData{
+			Operator: od,
+			BaseData: ab.BaseData{
+				Token: string(v),
+			},
+		}
+		ac, err := jwt.DecodeAccountClaims(a.Token)
+		if err != nil {
+			return err
+		}
+		a.Claim = ac
+		a.Loaded = a.Claim.IssuedAt
+		a.EntityName = a.Claim.Name
+		a.Key, err = p.getKey(a.Claim.Subject)
+		if err != nil {
+			return err
+		}
+		for pk := range a.Claim.SigningKeys {
+			k, err := p.getKey(pk)
+			if err != nil {
+				return err
+			}
+			a.AccountSigningKeys = append(a.AccountSigningKeys, k)
+		}
+		if err := p.loadUsers(a); err != nil {
+			return err
+		}
+		od.AccountDatas = append(od.AccountDatas, a)
+	}
+	return nil
+}
+
+func (p *Provider) loadUsers(ad *ab.AccountData) error {
+	// users stored under <accountPublicKey>.<userPublicKey>
```

Comment:
be good to get a little README or something that pulls this design out into a single place